### DISCUSSION TITLE
Redirecting 404s from Livewire routes inside Exception handler produces error "Return value must be of type Symfony\Component\HttpFoundation\Response, Livewire\Features\SupportRedirects\Redirector returned"

### DISCUSSION THREAD

**Discussion by ralphmorris** [livewire/livewire#8857] [7. Bugs (archived)]
Hey,

I have a number of routes going straight to Livewire components. Example:

```
Route::get('/resource/{slug}', ResourceShow::class)->name('resource.show');
```

When {slug} isn't found it triggers a 404 response.

I have created a Redirect feature that catches NotFoundHttpException and ModelNotFoundException and checks to see if there have been any redirects put in place before serving a 404 error. Code below:

```
public function render($request, Throwable $e)
{
    if ($e instanceof NotFoundHttpException || $e instanceof ModelNotFoundException) {
        if ($redirect = Redirect::where('from', '/'.$request->path())->first()) {
            return redirect()->to($redirect->to, $redirect->status_code);
        }
    }
    return parent::render($request, $e);
}
```

This works normally on non Livewire routes, however produces the following error due to Livewire overriding the default Redirect facade:

```
App\Http\Middleware\AcceptedTermsMiddleware::handle(): Return value must be of type Symfony\Component\HttpFoundation\Response, Livewire\Features\SupportRedirects\Redirector returned
```

I'm not sure if this is a bug or not, but it feels like it should be possible to have a single catch all place in the Exception handler to catch 404s and be able to redirect if necessary, rather than adding missing() to every route.

How can I redirect as per the above from the Exception Handler when in a Livewire page request?

Thanks!

Ralph

**Comment by EdgarsJoja** [comment]
I had similar issue, albeit not in a middleware. In the end I just removed type-hint from the method.
But I guess, if you still want to keep types, then using union might work:
```php
public function handle(...): SymfonyResponse | LivewireRedirector
```

Unless Laravel has some type expectations for `handle` methods, this should work.

**Reply by ralphmorris** [reply]
I tried removing the type hints but unfortunately caused another error further down the stack.

```
Undefined property: Livewire\Features\SupportRedirects\Redirector::$headers
```

I guess that Livewire requires the response in one format while Laravel requires a slightly different format. I think using redirect() inside of an exception handler is pretty common so I'm surprised I can't find more posts about this online.

**Reply by patoui** [reply]
@ralphmorris Currently experiencing this same issue, have you found a solution?

**Reply by ralphmorris** [reply]
@patoui I ended up redirecting within the mount method of the component. This can be wrapped up though to be more reuseable. 

Essentially don't type hint the model in the mount method and instead just pass it as a variable. Then do the query yourself with a find() instead of findOrfail(), then redirect if null.

Basic example below:

```
public function mount($slug = null)
{
    $this->page = PageModel::findPageBySlug($slug)->first();

    if (!$this->page) {
        return redirect('/');
    }
}
```

**Comment by thierrymaasdam** [comment]
Experiencing the same issue again, in the exact same scenario. Upvoted.
